lib/compress/zstd_lazy.c

```diff
@@ -475,6 +476,122 @@ U32 ZSTD_insertAndFindFirstIndex(ZSTD_matchState_t* ms, const BYTE* ip) {
     return ZSTD_insertAndFindFirstIndex_internal(ms, cParams, ip, ms->cParams.minMatch);
 }
 
+void ZSTD_dedicatedDictSearch_lazy_loadDictionary(ZSTD_matchState_t* ms, const BYTE* const ip)
```

Comment:
The DDS has its own indexing scheme. I wonder if we can add `chainSize + 1` to the index, so we can distinguish chain table indices from source indices. That would allow us to fit 4-entry chains into the hash table without spilling over.

1. Index `chainSize` (or 0, whichever is faster) means no chain - we have reached the end of our chain
2. Indices `[0, chainSize)` mean this is where the chain starts. We necessarily need a size, because we could `0` terminate the chain. Generally searching for a `0` element is faster than iterating over a size.
3. Any index in the hash table `>= chainSize + 1` is a valid index. If this is the 4th element in the hash table, then there is no chain.

Or, I wonder if we could fit both a chain index and a position in the hash table. If we limit dedicatedDictSearch to dictionaries <= 128KB we have 15 bits leftover for the chain table index. That means we can fit 32K entries in our chain table. That would allow us to utilize all 4 entries in our hash table. Then the first entry who has non-zero high 15 bits is the last entry in the hash table, and that is the position in the chain table.

I'm not sure if either is worth the effort. But something along these lines might pan out and be worth trying.